byterun/domain.c

```diff
@@ -300,19 +299,61 @@ static void* domain_thread_func(void* v) {
   return 0;
 }
 
+struct domain_thread {
+    caml_plat_mutex m;
+    pthread_t th;
+    int joinable;
+};
+
+#define Domainthreadptr_val(val) ((struct domain_thread**)Data_custom_val(val))
+
+static void domain_finalize(value v)
+{
+    struct domain_thread** dtp = Domainthreadptr_val(v);
+    struct domain_thread* dt;
+    Assert(dtp);
+    dt = *dtp;
+    Assert(dt);
+    caml_plat_mutex_free(&dt->m);
+    if (dt->joinable) {
+        caml_gc_log("Detaching thread");
+        pthread_detach(dt->th);
+    }
+    caml_stat_free(dt);
+    *dtp = NULL;
+}
+
+CAMLexport const struct custom_operations domain_ops = {
+    "domain",
+    domain_finalize,
+    custom_compare_default,
+    custom_hash_default,
+    custom_serialize_default,
+    custom_deserialize_default,
+    custom_compare_ext_default
+};
 
+CAMLprim value caml_ml_domain_join(value domain);
 CAMLprim value caml_domain_spawn(value callback)
 {
   CAMLparam1 (callback);
+  CAMLlocal1 (th_val);
   struct domain_startup_params p;
-  pthread_t th;
+  struct domain_thread **dp;
+  struct domain_thread *d;
   int err;
+  
+  th_val = caml_alloc_custom(&domain_ops, sizeof(*dp), 0, 1);
+  dp = Domainthreadptr_val(th_val);
+  *dp = d = caml_stat_alloc(sizeof(*d));
+  caml_plat_mutex_init(&d->m);
+  d->joinable = 1;
 
   caml_plat_event_init(&p.ev);
 
   p.callback = caml_create_root(caml_promote(&domain_self->state, callback));
 
-  err = pthread_create(&th, 0, domain_thread_func, (void*)&p);
+  err = pthread_create(&d->th, 0, domain_thread_func, (void*)&p);
```

Comment:
If this returns an error, then `d->th` has not been initialised. The finaliser will run, and since `joinable=true`, it will try to `pthread_detach(d->th)`.

I think `joinable` should be initialised to `0`, and only set to `1` if `pthread_create` succeeds.